### PR TITLE
Add `ContextKeyService`: a class to manage context values scoped on objects.  (used to eval expressions)

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This PR works with #3350 to provide a "context service" that manages `Contexts` (essentially just dicts that have parents, and pull from the parent when a key is missing).  This pattern is inspired by looking through the way that VScode manages the context objects that ["when clauses"](https://code.visualstudio.com/api/references/when-clause-contexts) can later evaluate.
The core idea here is that when you create an object (like a viewer or a layerlist, etc...) it gets its own "scoped" context service that looks in higher up parent contexts when a key is missing, but can locally override values (much like python scopes)... all the way up the root context, which can look in `settings` for values.

This way, there is no "global" context that has to get updated as we swap viewers.  There _is_ a root context, but each viewer has it's own scoped sub-context, and each layerlist is sub-scoped off of it corresponding viewer, etc...

```python
from napari.utils.context._service import get_context, create_context

class A:
    def __init__(self) -> None:
        self.ctx = create_context(self)
        self.b = B()  # self.b.ctx will be scoped off of self.ctx (using stack inspection)

class B:
    def __init__(self) -> None:
        self.ctx = create_context(self)

obj = A()
assert get_context(obj) is obj.ctx
```

updating contexts...

```py
In [2]: obj.ctx['a'] = 1

In [3]: obj.b.ctx['a']
Out[3]: 1

In [4]: obj.b.ctx['a'] = 3

In [5]: obj.b.ctx['a']
Out[5]: 3

In [6]: obj.ctx['a']
Out[6]: 1

In [7]: obj.ctx['settings.appearance.theme']
Out[7]: 'dark'
```

evaluating expressions with context:

```python
In [8]: from napari.utils.context._expressions import parse_expression

In [9]: expr = parse_expression('a > 4')

In [10]: expr = parse_expression('a > 2')

In [11]: expr.eval(obj.ctx)
Out[11]: False

In [12]: expr.eval(obj.b.ctx)
Out[12]: True
```

for now, I have left out the "hookup" to the layerlist, but ~will open~ have opened  a follow up PR (#3376) that explains how A/B would update their contexts and how this all fits together (but figured it's maybe easier to review and verify testing of this more abstract part on it's own)

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by goanpeca** [comment]
This is pretty cool @tlambert03 🚀 !

File: napari/utils/context/_service.py
Comment (Carreau) [inline napari/utils/context/_service.py:48]: I'm commenting as I'm reading so it might be answered later. 

Why not use ChainMap ? https://docs.python.org/3/library/collections.html#collections.ChainMap

Comment (tlambert03) [reply napari/utils/context/_service.py:48]: oh hell yes... thank you for reminding me.

File: napari/utils/context/_tests/test_expressions.py
Comment (Carreau) [inline napari/utils/context/_tests/test_expressions.py:219]: Jut to make sure I understand, 

Here you are evaluating in the root context, but you should be able to evaluate in the any scope context ? Is that right ?  Do you want actually evaluate in a scoped context ? 

I'm not too sure I see yet how this will interact with multiple objects and `when` clauses, in particular when you have many instances for examples of layers, but I guess you know where you are going, and in particular why you `root.get_context(obj)`, and not `root.get_context(key)`, I I would have thought that in expressions like `layer.active.n_dims > 2` for example, there would be a `root.get_context('layer').get_context('active').get_context('n_dims')`. But i'm likely misunderstanding how this will be used.

File: napari/utils/context/_service.py
Comment (Carreau) [inline napari/utils/context/_service.py:133]: Still Todo. Is that Ok ?

File: napari/utils/context/_tests/test_expressions.py
Comment (tlambert03) [reply napari/utils/context/_tests/test_expressions.py:219]: > Here you are evaluating in the root context, but you should be able to evaluate in the any scope context ? Is that right ? Do you want actually evaluate in a scoped context ?

yes that's correct.  for the purpose of this test, I'm mostly making sure that I can evaluate differently.  normally we'd be evaluating in a scoped context I assume.

> But i'm likely misunderstanding how this will be used.

No, I think your question is a really good one.  For the moment, I was just focusing on the concept of a scoped context. but _exactly_ how we pick a context to evaluate is still somewhat of a question in my mind.  I guess, where a viewer has a number of singleton objects (like a single layerlist for instance), we could easily merge specific contexts prior to evaluation.  but that is indeed an open question

File: napari/utils/context/_service.py
Comment (sofroniewn) [reply napari/utils/context/_service.py:48]: Is this something you want to do before merge?

**Review by sofroniewn — APPROVED** [review]
I believe both @jni owed you a review on this. Changes look good to me. Sorry they sat here so long. Is there anything you still want to do before merge? Maybe @jni can take a quick look and then we can get this in and unblock you!!

File: napari/utils/context/_service.py
Comment (jni) [inline napari/utils/context/_service.py:27]: Make this a permalink?

Comment (jni) [reply napari/utils/context/_service.py:27]: Or link to a version tag?

File: napari/utils/context/_service.py
Comment (jni) [reply napari/utils/context/_service.py:48]: :+1:!

File: napari/utils/context/_service.py
Comment (jni) [inline napari/utils/context/_service.py:94]: Can you add a docstring here detailing the logic of the function? I guess it allows things like ctx['a.b.c']? What does it mean, though? I don't quite follow the `getattr(val, ...)` and `val.dict()`. A docstring could clarify the whole thing.

File: napari/utils/context/_service.py
Comment (jni) [inline napari/utils/context/_service.py:23]: Can you add a further example demonstrating the nesting of scopes? Currently it's unclear why there's a class B in this example. :joy:

File: napari/utils/context/_service.py
Comment (jni) [inline napari/utils/context/_service.py:81]: Can you add a full docstring including usage examples?

File: napari/utils/context/_service.py
Comment (jni) [reply napari/utils/context/_service.py:23]: The more I read the more I'm confused about the terminology of "Service" and "Key" and "KeyService"... Perhaps you could also expand on that in this docstring?

File: napari/utils/context/_service.py
Comment (jni) [reply napari/utils/context/_service.py:133]: I am also wondering this.

File: napari/utils/context/_service.py
Comment (jni) [inline napari/utils/context/_service.py:118]: Even though this is a private class, it would be great to have docstrings for future visitors to the code. To be honest the whole PR could benefit from more docstrings!

File: napari/utils/context/_service.py
Comment (jni) [inline napari/utils/context/_service.py:320]: Why not a for-loop? (with `if not frame: ...` inside)

File: napari/utils/context/_service.py
Comment (jni) [inline napari/utils/context/_service.py:301]: Should `depth` be renamed `max_depth`?

**Review by jni — COMMENTED** [review]
Thanks for this @tlambert03! tbh I would very much like very many more comments/docstrings in this code, as I found it difficult (ironically) to read the code without, erm, context. :joy: I think that moving a few of the tests to be doctests, for example, would be useful.

**Comment by tlambert03** [comment]
thank you all for the reviews, very helpful.  I'm closing this in favor of a much-simplified context object, and a PR merged with #3376 to make it all easier to appreciate in "context" ;).

https://github.com/napari/napari/pull/3513

**Pull request closed without merging**
